### PR TITLE
Implement failures.c

### LINKED ISSUE #67: Implement failures.c

### PR DESCRIPTION
**Pull request by Tsjerfell**
Fixes #67

### CODE REVIEW

File: Nouveau PIR/firmware/source/failures.c [hunk 1]
```diff
@@ -535,24 +535,24 @@
 	{	
 		if(CRK_signal == false)
 		{
-			TIM_Cmd(TIM1, ENABLE);//Enable timer 6 (notre timer 1)
+			TIM_Cmd(TIM1, ENABLE);  //Enable timer1 (formerly Timer6 on microchip)
```
Comment (hmathieu-insat) [inline Nouveau PIR/firmware/source/failures.c:538]: Why?
On the google sheet, timer6 on microchip is said to correspond with Timer3

File: Nouveau PIR/firmware/source/failures.c [hunk 2]
```diff
@@ -658,11 +659,10 @@
 //## CAM_delay: CAM_TOOTH_OFF / CAM_REF_CRK / CAM_SYN / CAM_SYN_CRK
 	void CAM_delay (int cam_id)
 	{
-		if(TIM4->CR1 & TIM_CR1_CEN)
+		if(TIM4->CR1 & TIM_CR1_CEN) // if timer4 is enabled
 		{
```
[the remark this answers is not in the document]
Comment (hmathieu-insat) [reply Nouveau PIR/firmware/source/failures.c:663]: (The one based on our Timer1)

File: Nouveau PIR/firmware/source/failures.c [hunk 3]
```diff
@@ -699,7 +699,7 @@
 
 			if(((double)SysTick->VAL/former_teeth_time) * revolution_CRK >= (delay_angle_CAM_delay * delay_factor_CAM_delay))
 			{
-				TIM_Cmd(TIM8,DISABLE);
+				TIM_Cmd(TIM8,DISABLE); //TODO: #77 Port with systick
```
Comment (hmathieu-insat) [inline Nouveau PIR/firmware/source/failures.c:702]: This should be ported with the systick (we have no Timer8 on the microchip)
[#77](https://github.com/hmathieu-insat/PIR/issues/77)

File: Nouveau PIR/firmware/source/failures.c [hunk 4]
```diff
@@ -749,7 +749,7 @@
 
 				if(shift_counter_CAM_delay[cam_id][i] != 0)
 				{
-					if(angle_to_edge_CAM_delay[cam_id][i] + ((double)(shift_counter_CAM_delay[cam_id][i] - 1) + ((double)((unsigned long)TMR2 + timer_overflow_CRK*(unsigned long)PR2))/former_teeth_time) * revolution_CRK >= (delay_angle_CAM_delay * delay_factor_CAM_delay))
+					if(angle_to_edge_CAM_delay[cam_id][i] + ((double)(shift_counter_CAM_delay[cam_id][i] - 1) + ((double)((unsigned long)TIMGetCounter(TIM4) + timer_overflow_CRK*(unsigned long)(TIM2->ARR)))/former_teeth_time) * revolution_CRK >= (delay_angle_CAM_delay * delay_factor_CAM_delay)) //Aurait été plus propre avec un getAutoreload
```
Comment (hmathieu-insat) [inline Nouveau PIR/firmware/source/failures.c:752]: Shouldn't the software based timer implement this?

**Review by hmathieu-insat — CHANGES_REQUESTED** [review]
Did you implement the software timer, I couldn't find the corresponding commit if you did.
Some changes to tend to I think but shouldn't take long.

File: Nouveau PIR/firmware/source/failures.c (hunk 1 above)
Comment (Tsjerfell) [reply Nouveau PIR/firmware/source/failures.c:538]: You're absolutely right

File: Nouveau PIR/firmware/source/failures.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tsjerfell) [reply Nouveau PIR/firmware/source/failures.c:663]: It indeed should

File: Nouveau PIR/firmware/source/failures.c [hunk 5]
```diff
@@ -882,13 +888,10 @@
...
-		T8CONbits.TON = 0;
-		TMR8 = 0;
-		T9CONbits.TON = 0;
-		Timer9Reset();
-		T4CONbits.TON = 0;
-		Timer4Reset();		
-		number_processing_edges_CAM_delay[0] = 0;
+		SysTick->CTRL &= ~SysTick_CTRL_ENABLE; // disable SysTick
+		SysTick->VAL = (2^24)-1;  //clear systick counter
+		//TODO: #78 There is a deactivation + reset of former timer4 (new software implemented timer) to implement here
```
Comment (hmathieu-insat) [inline Nouveau PIR/firmware/source/failures.c:918]: TODO has not been addreesed here unless I'm mistaken

**Review by hmathieu-insat — CHANGES_REQUESTED** [review]
Apart from 1 remark, seems all good to me

**Review by hmathieu-insat — APPROVED** [review]

**Pull request merged**